File: python/00_common.py
"""
00_common.py

Project : Re-analysis of the NCCTG Adjuvant Colon Cancer Randomized Trial
          with Frequentist, Bayesian, and Machine Learning Methods
Author  : Cathy
Purpose : Shared configuration for the Python analysis modules: paths,
          the Okabe-Ito colour-blind-safe palette, plotting style, and
          input/output helpers. Mirrors R/00_common.R for a coherent visual
          identity across the two languages.
"""

import os
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.ticker as mtick

# ---- Paths --------------------------------------------------------------------
FIG_DIR = "outputs/figures"
TAB_DIR = "outputs/tables"
DATA_PROC = "data/processed"
os.makedirs(FIG_DIR, exist_ok=True)
os.makedirs(TAB_DIR, exist_ok=True)

# ---- Trial arm configuration (matches R/00_common.R) ---------------------------
ARM_LEVELS = ["Obs", "Lev", "Lev+5FU"]
ARM_LABELS = {
    "Obs": "Observation",
    "Lev": "Levamisole",
    "Lev+5FU": "Levamisole + 5-FU",
}
ARM_COLS = {
    "Obs": "#0072B2",
    "Lev": "#E69F00",
    "Lev+5FU": "#009E73",
}
TREATMENT_COLS = {"control": "#454545", "active": "#B22222"}
ACCENT = "#B22222"
BLUE = "#0072B2"
GREEN = "#009E73"
ORANGE = "#E69F00"
GREY = "#454545"

# ---- Global seed and version reporting ----------------------------------------
RANDOM_SEED = 20260916


def set_seed(seed: int = RANDOM_SEED) -> None:
    import random
    random.seed(seed)
    np.random.seed(seed)


# ---- Plotting style ------------------------------------------------------------
def apply_style() -> None:
    """Apply a clean, academic ggplot-like style consistent with the R figures."""
    plt.rcParams.update({
        "figure.facecolor": "white",
        "axes.facecolor": "white",
        "axes.spines.top": False,
        "axes.spines.right": False,
        "axes.grid": True,
        "grid.color": "#e2e2e2",
        "grid.linewidth": 0.6,
        "axes.edgecolor": "#444444",
        "axes.linewidth": 0.8,
        "font.size": 11,
        "axes.titlesize": 12,
        "axes.titleweight": "bold",
        "axes.labelsize": 11,
        "legend.frameon": False,
        "legend.fontsize": 10,
        "xtick.labelsize": 10,
        "ytick.labelsize": 10,
        "savefig.dpi": 300,
        "savefig.bbox": "tight",
        "figure.autolayout": True,
    })


apply_style()


def save_fig(fig, name: str) -> None:
    """Save a matplotlib figure as PNG (README) and PDF (report)."""
    fig.savefig(os.path.join(FIG_DIR, f"{name}.png"), facecolor="white")
    fig.savefig(os.path.join(FIG_DIR, f"{name}.pdf"), facecolor="white")
    print(f"  figure saved: {FIG_DIR}/{name}.png")
    plt.close(fig)


def save_tab(df: pd.DataFrame, name: str) -> None:
    """Save a tidy data frame as CSV in outputs/tables."""
    path = os.path.join(TAB_DIR, f"{name}.csv")
    df.to_csv(path, index=False)
    print(f"  table saved: {path}")


def load_analysis_data() -> pd.DataFrame:
    """Load the patient-level analysis file produced by R/01_data_curation.R."""
    df = pd.read_csv(os.path.join(DATA_PROC, "colon_analysis.csv"))
    df["rx"] = pd.Categorical(df["rx"], categories=ARM_LEVELS)
    return df


def design_matrix(df: pd.DataFrame, drop_reference: bool = True) -> pd.DataFrame:
    """Build the modelling design matrix with clinically labelled columns."""
    X = pd.DataFrame(index=df.index)
    X["age"] = df["age"]
    X["nodes"] = df["nodes"]
    X["female"] = (df["sex"] == "Female").astype(int)
    X["obstruction"] = (df["obstruct"] == "Yes").astype(int)
    X["perforation"] = (df["perfor"] == "Yes").astype(int)
    X["adherence"] = (df["adhere"] == "Yes").astype(int)
    X["diff_moderate"] = (df["differ"] == "Moderate").astype(int)
    X["diff_poor"] = (df["differ"] == "Poor").astype(int)
    X["extent_muscle"] = (df["extent"] == "Muscle").astype(int)
    X["extent_serosa"] = (df["extent"] == "Serosa").astype(int)
    X["extent_contiguous"] = (df["extent"] == "Contiguous structures").astype(int)
    X["surgery_long"] = (df["surg"] == "Long").astype(int)
    return X.astype(float)


def fmt_p(p: float) -> str:
    if np.isnan(p):
        return "NA"
    if p < 0.001:
        return "< 0.001"
    return f"{p:.3f}" if p < 0.01 else f"{p:.2f}"


def percent_axis(ax, axis: str = "y") -> None:
    if axis == "y":
        ax.yaxis.set_major_formatter(mtick.PercentFormatter(1.0))
    else:
        ax.xaxis.set_major_formatter(mtick.PercentFormatter(1.0))

File: R/00_common.R
# ==============================================================================
# 00_common.R
#
# Project : Re-analysis of the NCCTG Adjuvant Colon Cancer Randomized Trial
#           with Frequentist, Bayesian, and Machine Learning Methods
# Author  : Cathy
# Purpose : Shared configuration: paths, labels, colour palette, plotting
#           theme, and helper functions used by all analysis scripts.
#
# Note    : All scripts are intended to be run from the repository root, e.g.
#             Rscript R/01_data_curation.R
# ==============================================================================

suppressPackageStartupMessages({
  library(dplyr)
  library(tidyr)
  library(purrr)
  library(ggplot2)
  library(patchwork)
  library(survival)
})

# ---- Paths -------------------------------------------------------------------
FIG_DIR <- "outputs/figures"
TAB_DIR <- "outputs/tables"
DATA_RAW <- "data/raw"
DATA_PROC <- "data/processed"
dir.create(FIG_DIR, showWarnings = FALSE, recursive = TRUE)
dir.create(TAB_DIR, showWarnings = FALSE, recursive = TRUE)

# ---- Trial arm configuration -------------------------------------------------
ARM_LEVELS <- c("Obs", "Lev", "Lev+5FU")
ARM_LABELS <- c(
  "Obs"     = "Observation",
  "Lev"     = "Levamisole",
  "Lev+5FU" = "Levamisole + 5-FU"
)
# Okabe-Ito colour-blind-safe palette
ARM_COLS <- c(
  "Obs"     = "#0072B2",
  "Lev"     = "#E69F00",
  "Lev+5FU" = "#009E73"
)
TREATMENT_COLS <- c("#454545", "#B22222")  # control / active in two-group plots

# ---- Variable labels ---------------------------------------------------------
VAR_LABELS <- c(
  rx      = "Treatment arm",
  sex     = "Sex",
  age     = "Age (years)",
  obstruct= "Obstruction",
  perfor  = "Perforation",
  adhere  = "Adherence to nearby organs",
  nodes   = "Positive lymph nodes",
  differ  = "Tumour differentiation",
  extent  = "Extent of local spread",
  surg    = "Timing of surgery to registration",
  node4   = ">= 4 positive nodes",
  time    = "Time (days)",
  status  = "Event indicator"
)

# ---- ggplot theme (academic, colour-blind safe) ------------------------------
theme_trial <- theme_bw(base_size = 11) +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_line(linewidth = 0.25, colour = "grey88"),
    strip.background = element_rect(fill = "grey95", colour = NA),
    strip.text = element_text(face = "bold", size = 10),
    legend.position = "bottom",
    legend.title = element_blank(),
    axis.title = element_text(size = 10),
    plot.title = element_text(size = 11, face = "bold"),
    plot.subtitle = element_text(size = 9, colour = "grey30"),
    plot.caption = element_text(size = 7.5, colour = "grey40")
  )

# ---- Helpers -----------------------------------------------------------------
`%||%` <- function(a, b) if (is.null(a)) b else a

# Save a ggplot object as both PNG (for README) and PDF (for the report).
save_fig <- function(plot, name, width = 7, height = 5, dpi = 300) {
  stopifnot(grepl("\\.(png|pdf)$", name)[1] || TRUE)
  png_path <- file.path(FIG_DIR, paste0(name, ".png"))
  pdf_path <- file.path(FIG_DIR, paste0(name, ".pdf"))
  ggsave(png_path, plot, width = width, height = height, dpi = dpi, bg = "white")
  ggsave(pdf_path, plot, width = width, height = height, bg = "white")
  message("  figure saved: ", png_path)
  invisible(plot)
}

# Write a tidy data frame as CSV in outputs/tables.
save_tab <- function(df, name) {
  path <- file.path(TAB_DIR, paste0(name, ".csv"))
  write.csv(df, path, row.names = FALSE, na = "")
  message("  table saved: ", path)
  invisible(df)
}

# Round to fixed digits while keeping numeric class.
rounddf <- function(df, digits = 3) {
  df[] <- lapply(df, function(x) if (is.numeric(x)) round(x, digits) else x)
  df
}

# Format p-values in APA-ish style.
fmt_p <- function(p) {
  ifelse(is.na(p), NA_character_,
  ifelse(p < 0.001, "< 0.001",
  ifelse(p < 0.01, sprintf("%.3f", p),
  sprintf("%.2f", p))))
}

# ---- Reproducibility ---------------------------------------------------------
set_global_seed <- function(seed = 20260916) {
  set.seed(seed)
  invisible(seed)
}
